lib/srv/desktop/tdp/proto.go

```diff
@@ -1504,3 +1395,35 @@ func decodeString(r io.Reader, maxLen uint32) (string, error) {
 	}
 	return string(s), nil
 }
+
+// rawMsg is a tdp.Message that has not been decoded.
+// Its encode operation is therefore a no-op.
+type rawMsg []byte
+
+func (r rawMsg) Encode() ([]byte, error) { return []byte(r), nil }
+
+// writeUint16 writes v to b in big endian order
+func writeUint16(b *bytes.Buffer, v uint16) {
+	b.WriteByte(byte(v >> 8))
+	b.WriteByte(byte(v))
+}
+
+// writeUint32 writes v to b in big endian order
+func writeUint32(b *bytes.Buffer, v uint32) {
+	b.WriteByte(byte(v >> 24))
+	b.WriteByte(byte(v >> 16))
+	b.WriteByte(byte(v >> 8))
+	b.WriteByte(byte(v))
+}
+
+// writeUint64 writes v to b in big endian order
+func writeUint64(b *bytes.Buffer, v uint64) {
+	b.WriteByte(byte(v >> 56))
+	b.WriteByte(byte(v >> 48))
+	b.WriteByte(byte(v >> 40))
+	b.WriteByte(byte(v >> 32))
+	b.WriteByte(byte(v >> 24))
+	b.WriteByte(byte(v >> 16))
+	b.WriteByte(byte(v >> 8))
+	b.WriteByte(byte(v))
+}
```

Comment:
Are these more performant?

Reply:
Details in the commit message, but yes. `binary.Write` accepts an `interface{}` and uses reflection to determine the type (int32, uint64, etc.).  Reflection is quite a bit slower than just writing bytes directly.

In fact, Go 1.19 introduced new functions that do exactly this, but they work on `[]byte` instead of `bytes.Buffer` (see `binary.BigEndian.AppendUint16()`, for example)